### ISSUE TITLE
Supporting Gurobi Instant Cloud

### ISSUE THREAD

**Issue by c-kob** [optimatika/ojAlgo-gurobi#5]
The way in which the delegate solver is set up does not appear to allow the use of Gurobi Instant Cloud services but requires the optimiser to reside on the machine itself.

It should be straightforward to provide an alternative constructor that takes as parameters a CloudAccessID and a CloudSecretKey (e.g. SolverGurobi.withInstantCloud(accessKey, secret) as an alternative to SolverGurobi.INTEGRATION) that sets up a GRBEnv for use with a cloud instance, e.g.

```java
 Integration(final String accessKey, final String secret) {
            super();
            GRBEnv tmpGRBEnv;
            try {
                if (accessKey != null && secret != null) {
                    tmpGRBEnv = new GRBEnv(null, accessKey, secret, null, 0);
                } else {
                    tmpGRBEnv = new GRBEnv();
                }
            } catch (final GRBException anException) {
                tmpGRBEnv = null;
            }
            myEnvironment = tmpGRBEnv;
        }
```

This can then be used as:

```java
 public static final GurobiSolver.Integration INTEGRATION = new Integration(null, null);

    public static GurobiSolver.Integration withInstantCloud(final String accessKey, final String secret) {
        return new Integration(accessKey, secret);
    }
```

With instant cloud use, it might also be preferable to avoide the use of finalize() and implement AutoCloseable instead.

**Comment by apete** [comment]
I created a branch https://github.com/optimatika/ojAlgo-gurobi/tree/Supporting-Gurobi-Instant-Cloud and implemented your suggestions.

It's been a while since I used this solver integration. Currently do not have the ability to use/test Gurobi. (This doesn't even compile with my version of gurobi.jar.)

**Comment by c-kob** [comment]
That's great - thanks. 
Happy to fork and create a pull request if that's helpful.
I made those changes locally and it seems to work fine (incidentially, I couldn't run with the dependendy on ojAlgo-gurobi as there was an ClassNotFound error complaining about not finding PrimitiveMath 
(java.lang.NoClassDefFoundError: org/ojalgo/constant/PrimitiveMath
	at org.ojalgo.optimisation.solver.gurobi.SolverGurobi$Integration.build(SolverGurobi.java:110) ~[ojalgo-gurobi-2.47.2.jar!/:2.47.2])

**Comment by apete** [comment]
Please fork, test and create a PR when you're satisfied.

The version that is published to Maven Central still depends on an older version of ojAlgo. Need to fix that.

**Comment by c-kob** [comment]
I've forked and updated and am reasonably happy with the way the integration works. I've also tried to adjust and run the tests that are included. They don't all pass, but that doesn't seem to have anything to do with using the cloud solver instead of a binary running on the local machine, but probably with having to tweak the Gurobi options. Specifically, the test for GitHubIssue2 in the ExtensionsSolveProblems Suite fails (AssertionFailedError: expected: <25.8> but was: <16.6>) and 31 of the Netlib tests up to KEN_18 fail - KEN_18 doesn't seem to take back an answer even though just running the MPS through Gurobi works fine). 
Not sure what to make of that ...

**Comment by apete** [comment]
That GitHubIssue2 test case is not good since it's an unbounded problem. The test case seems to check for the feasible solution ojAlgo returns. (Should probably change that test.) Running with CPLEX I get this solution:

INVALID 16.6 @ { 10, 0, 8, 0, 5, 0 }

It would be much appreciated if you could help figure out why so many go Netlib cases fail.

One thing to test is if the ojAlgo presolvers are breaking things. Call `ExpressionsBasedModel.clearPresolvers()` to see if that changes anything.

Or if it's possible to compare the problem instances depending on if Gurobli read the MPS directly or if it passed through ojAlgo.

**Comment by c-kob** [comment]
At least CPLEX and Gurobi seem to agree on 16.6 ...

I'll have a look over the next few days if possible. Will keep you posted.

**Comment by apete** [comment]
For `NetlibCase` (`NetlibCaseWithGurobi`) the individual tests may be tagged with `slow` and/or `unstable`.

You should be able to include/exclude tests based on these tags. The subset of tests that are neither tagged `slow` nor `unstable` definitely should work. That subset works both with the ojAlgo solvers and with CPLEX (verified now). Definitely should work with GUROBI as well.

**Comment by c-kob** [comment]
I've create a pull request. Will look into the failing tests with the slow/unstable tag when I get better access to a machine with a working local Gurobi install (the problems don't seem to be related to where Gurobi runs).

**Comment by apete** [comment]
Thank you. I accepted that PR, as is, but I think maybe you didn't get som of the changes I had already done in that branch. I'll review and possibly change some things. Make sure to rebase before creating another PR.

**Comment by c-kob** [comment]
Apologies - I thought I had :-(

**Comment by apete** [comment]
No problem, it was just some very minor things that got lost. I pushed an update now. It's mostly "formatting". Hope that doesn't upset you.

I did change the ojAlgo dependency to `48.4.1-SNAPSHOT` from the `Update-external-solvers` branch. Ok for you to build from source? The intention is to publish new versions of all 3:d party solver libs.

**Comment by c-kob** [comment]
Great - and no worries, I'm not easily upset ;-)
Building from source is fine with me.
Thanks for all the work, by the way.

**Comment by apete** [comment]
In the pom you put the `maven-surefire-plugin` in `<pluginManagement>` – what's the purpose/benefit of that? Are you using this pom as a parent? Or are these configurations also transferred to projects that depend on this?

**Comment by c-kob** [comment]
My maven complained when declaring the plugin directly in plugins rather than <pluginManagement>. 
And looking at the website suggested that including it in pluginManagement is correct (https://maven.apache.org/surefire/maven-surefire-plugin/usage.html <https://maven.apache.org/surefire/maven-surefire-plugin/usage.html>)

**Comment by apete** [comment]
That example describes usage in a parent pom.

**Comment by apete** [comment]
I've released ojAlgo v48.4.1 and ojAlgo-gurobi v3.0.0.

Let me know if there's anything that doesn't work, and close this issue if/when you're happy.

**Comment by c-kob** [comment]
Great - thanks!
